### ISSUE TITLE
Change capture FPS

### ISSUE THREAD

**Issue by dafreak29** [Genymobile/scrcpy#488]
Could you add a command line to change the capture fps (i.e. to 30fps)? I believe it is locked/capped at 60fps at the moment.

**Comment by rom1v** [comment]
I already tried some time ago, by requesting a specific FPS [here](https://github.com/Genymobile/scrcpy/blob/v1.8/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java#L141-L147), but it has no effect: the encoder takes one frame on every surface update (it's not a constant FPS).

**Comment by mtvg** [comment]
This would be an awesome feature, especially for over wifi capture. Slowing down to even lower fps would be great. 
Is there any way to skip frames before it reaches the encoder, based on timestamp? So If the SurfaceFlinger is sending 60 frames in 1 second, only send a few to the encoder (making sure the last frame is sent after some timeout if no new frame is coming)

**Comment by rom1v** [comment]
This is supported since Android Q ([`KEY_MAX_FPS_TO_ENCODER`](https://developer.android.com/reference/android/media/MediaFormat.html#KEY_MAX_FPS_TO_ENCODER)):

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
index 52f6f26..a55e84a 100644
--- a/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
+++ b/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java
@@ -6,6 +6,7 @@ import android.graphics.Rect;
 import android.media.MediaCodec;
 import android.media.MediaCodecInfo;
 import android.media.MediaFormat;
+import android.os.Build;
 import android.os.IBinder;
 import android.os.Looper;
 import android.view.Surface;
@@ -157,6 +158,9 @@ public class ScreenEncoder implements Device.RotationListener {
         format.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, iFrameInterval);
         // display the very first frame, and recover from bad quality when no new frames
         format.setLong(MediaFormat.KEY_REPEAT_PREVIOUS_FRAME_AFTER, MICROSECONDS_IN_ONE_SECOND * REPEAT_FRAME_DELAY / frameRate); // µs
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            format.setFloat(MediaFormat.KEY_MAX_FPS_TO_ENCODER, 10f);
+        }
         return format;
     }
 ```

**Comment by mtvg** [comment]
🤯 Thank you! Works perfectly, exactly what I was looking for.

**Comment by rom1v** [comment]
Implemented `--max-fps` option on `dev`.

**Comment by shiwf** [comment]
I've tried to set KEY_MAX_FPS_TO_ENCODER in AndroidO(Samsung Galalxy S7) and P(Samsung Galalxy S8+), they all could work. Actually, this key was available since O, they make it public until Android10.

**Comment by rom1v** [comment]
@shiwf Ah yes, it works even on Android 6.

Thank you: 63caff822690332a169005208f2994f39d5a6e41

**Comment by vegaskukichyo** [comment]
Hi, thank you for this amazing tool. I use it for livestreaming mobile games. I do not have a high-end computer, so frames are skipped, but fps stays around 55-60. I stream at 30fps, so capping the output at 30fps would eliminate frame drop and improve the smoothness of my stream.

I am on Android 9 (OnePlus 7 Pro).  I saw your commit with the added max-fps command for Android <10. I do not know how to compile the program, however. I use your pre-built archive, but I would like this option. Do I need to learn how and recompile the program, or could you upload a modified executable with this option? My archive already contains all the dependencies, as with all of your prebuilt archives.

Thank you.

**Comment by rom1v** [comment]
> I do not know how to compile the program, however.

[BUILD](https://github.com/Genymobile/scrcpy/blob/master/BUILD.md)

> Do I need to learn how and recompile the program, or could you upload a modified executable/server with this option?

Sorry, I can't for now (holidays...).

**Comment by vegaskukichyo** [comment]
Thanks for your quick reply. It would be quite the effort and time for me to teach myself to build the program, even with your excellent instructions. I don't have much experience with any of that. 

I hope you have a great holiday! Do you intend to implement these changes in a new version sometime soon (in the new year)? I hope so.

If I could donate even a few dollars to support your efforts and your project, I would also like to do that. I know you've put a lot of work into this.

Again, happy holidays, and thank you for all your hard work!

**Comment by rom1v** [comment]
- [`scrcpy.exe`](https://tmp.rom1v.com/scrcpy/issue488/1/scrcpy.exe)  
_SHA256: a81ee325c6cd03ff83d58b5900a9a8286c1fbec8685f67210c4e6830c2878159_
- [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/issue488/1/scrcpy-server)  
_SHA256: 0ae5675a571e91a3d4ec900c51b29b15fcd5317cf76daf1ad875d4054c8f1a55_

**Comment by vegaskukichyo** [comment]
Awesome, thank you! I just had an update come in that brought me to Android 10, though. I have a secondary device that I use sometimes that is on Android 9, so I will install this too. Thanks!

I hope you had a good holiday!

**Comment by rom1v** [comment]
@shiwf @vegaskukichyo Released in v1.13.
